### ISSUE TITLE
jinja2.exceptions.UndefinedError: 'dict object' has no attribute 'cards'

### ISSUE THREAD

**Issue by tpvasconcelos** [danielfrg/jupyter-flex#92]
Hey @danielfrg !

Just tried flex for the first time. I was able to easily reproduce the Plotly example in the docs. However, when I tried applying it to an already existing notebook I got the following Jinja exception. Since you are the author of this tool, I think you can probably tell what went wrong here. Let me know if this is an easy fix as I was looking forward to applying this to an existing project, as a POC.

🚀🚀🚀

```diff
$ jupyter nbconvert --to flex eda.ipynb
[NbConvertApp] Converting notebook eda.ipynb to flex
Traceback (most recent call last):
  File "/opt/miniconda3/bin/jupyter-nbconvert", line 11, in <module>
    sys.exit(main())
  File "/opt/miniconda3/lib/python3.8/site-packages/jupyter_core/application.py", line 254, in launch_instance
    return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
  File "/opt/miniconda3/lib/python3.8/site-packages/traitlets/config/application.py", line 845, in launch_instance
    app.start()
  File "/opt/miniconda3/lib/python3.8/site-packages/nbconvert/nbconvertapp.py", line 350, in start
    self.convert_notebooks()
  File "/opt/miniconda3/lib/python3.8/site-packages/nbconvert/nbconvertapp.py", line 524, in convert_notebooks
    self.convert_single_notebook(notebook_filename)
  File "/opt/miniconda3/lib/python3.8/site-packages/nbconvert/nbconvertapp.py", line 489, in convert_single_notebook
    output, resources = self.export_single_notebook(notebook_filename, resources, input_buffer=input_buffer)
  File "/opt/miniconda3/lib/python3.8/site-packages/nbconvert/nbconvertapp.py", line 418, in export_single_notebook
    output, resources = self.exporter.from_filename(notebook_filename, resources=resources)
  File "/opt/miniconda3/lib/python3.8/site-packages/nbconvert/exporters/exporter.py", line 181, in from_filename
    return self.from_file(f, resources=resources, **kw)
  File "/opt/miniconda3/lib/python3.8/site-packages/nbconvert/exporters/exporter.py", line 199, in from_file
    return self.from_notebook_node(nbformat.read(file_stream, as_version=4), resources=resources, **kw)
  File "/opt/miniconda3/lib/python3.8/site-packages/nbconvert/exporters/html.py", line 119, in from_notebook_node
    return super().from_notebook_node(nb, resources, **kw)
  File "/opt/miniconda3/lib/python3.8/site-packages/nbconvert/exporters/templateexporter.py", line 384, in from_notebook_node
    output = self.template.render(nb=nb_copy, resources=resources)
  File "/opt/miniconda3/lib/python3.8/site-packages/jinja2/environment.py", line 1304, in render
    self.environment.handle_exception()
  File "/opt/miniconda3/lib/python3.8/site-packages/jinja2/environment.py", line 925, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "/opt/miniconda3/share/jupyter/nbconvert/templates/flex/index.html.j2", line 1, in top-level template code
    {%- import "flex.j2" as flex with context -%}
  File "/opt/miniconda3/lib/python3.8/site-packages/jinja2/environment.py", line 1412, in make_module
    return TemplateModule(self, ctx)
  File "/opt/miniconda3/lib/python3.8/site-packages/jinja2/environment.py", line 1542, in __init__
    body_stream = list(template.root_render_func(context))  # type: ignore
+  File "/opt/miniconda3/share/jupyter/nbconvert/templates/flex/flex.j2", line 238, in top-level template code
+    {%- set _ = vars.current_section["cards"].append(vars.current_card) -%}
  File "/opt/miniconda3/lib/python3.8/site-packages/jinja2/environment.py", line 474, in getattr
    return getattr(obj, attribute)
jinja2.exceptions.UndefinedError: 'dict object' has no attribute 'cards'
```

**Comment by danielfrg** [comment]
We are doing something wrong when parsing the notebook cells.
Could you upload the notebook somewhere?

**Comment by tpvasconcelos** [comment]
I could try coming up with a MWE that throws the same error

**Comment by danielfrg** [comment]
Yes, that would be ideal.

**Comment by danielfrg** [comment]
Leave the Markdown headers and the code cells that have tags (body) and just remove all the code from those and that would work.

**Comment by tpvasconcelos** [comment]
I can reproduce it when I add a Markdown header with no `body` cells coming after it. You can test it on the `.ipynb` bellow

```json
{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "19760b5c-08fd-414e-8cf5-5ecc47af3699",
   "metadata": {},
   "source": [
    "# Markdown header 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bc96208-c4c2-4459-ae6b-9ba9641f441f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.express as px\n",
    "import plotly.graph_objects as go"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c387c7c-5101-4c03-b891-db9b92870cc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "margin = go.layout.Margin(l=20, r=20, b=20, t=30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1d75821-8c43-45b5-b913-dbdb8729788c",
   "metadata": {
    "tags": [
     "body"
    ]
   },
   "outputs": [],
   "source": [
    "df = px.data.iris()\n",
    "fig = px.scatter(df, x=\"sepal_width\", y=\"sepal_length\")\n",
    "\n",
    "fig.update_layout(margin=margin)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a5bab20-c3a0-4337-9aca-dd6f3551d4d8",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-18T17:09:13.627360Z",
     "iopub.status.busy": "2021-09-18T17:09:13.627123Z",
     "iopub.status.idle": "2021-09-18T17:09:13.629817Z",
     "shell.execute_reply": "2021-09-18T17:09:13.629307Z",
     "shell.execute_reply.started": "2021-09-18T17:09:13.627337Z"
    }
   },
   "source": [
    "# Markdown header 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f04f8726-210a-4c03-8534-93213641af38",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "print(\"nothing happening here\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
```

What is the scope of this project? i.e. do you want to support any notebook structure as long as the user uses the `body` tags (and optional `### headers`)? This would be great as you can have a notebook with multiple analyses and plots, out of which you only want to use a few for the Flex dashboard.

**Comment by tpvasconcelos** [comment]
Just to clarify what I meant above.

I don't think that the error I'm reporting here is necessarily a bug since it is a side effect of a section of the notebook that has nothing to do with a flex report. This is fine if you want the project to only support notebook files purposely written to output a flex HTML file, and nothing else.

If I can chime in my personal take on this, I think that this might hinder the growth of the project since, in most instances, you will also want to share the notebook file / python code with other analysis or data scientists in your organization, while only sharing the flex dashboard with product or end-users. In most organizations, a review process is required before sharing a dashboard or an analytics report, in which case you will want to include other sections, comments, and possibly even extra plots in the notebook that do not make it into the final Flex report.

Nice tool btw! 💪

**Comment by danielfrg** [comment]
I totally agree with you.

The idea is to been able to generate a dashboard with as little changes as possible to an existing notebook thats why its only adding a couple of markdown headers and tags to the cells.

I do considir this a bug and will take a look as soon as i can.

**Comment by danielfrg** [comment]
This is fixed now on 0.7.2.
